## Chapter 2 - Exercise 2: Facebook Advertising Analysis - Part 2

#### CPA (Cost per acquisition hay Cost per action)
- Là chi phí nhà quảng cáo trả tiền cho mỗi hành động đặc biệt (như thanh toán, chấp nhận một đăng ký...) 
- CPA trên facebook: là số tiền trung bình mà bạn thanh toán cho mỗi hành động mà người dùng thực hiện đối với quảng cáo của bạn. Mức phí thanh toán chịu ảnh hưởng bởi đối tượng mà bạn nhắm mục tiêu và số nhà quảng cáo khác đang cạnh tranh để hiển thị quảng cáo của họ cho đối tượng của bạn. Ví dụ quảng cáo được thiết kế đẹp sẽ khuyến khích nhiều người thực hiện hành động và bạn càng nhận được nhiều hành động hơn cho ngân sách, chi phí cho mỗi hành động của bạn sẽ càng thấp. (source: [CPA])(https://www.thuatngumarketing.com/cost-per-action/)
![CPA](Data/CPA_1.jpg)

- Hoặc có thể tính theo **<font color="red">công thức cách 2</font>**: 
## $ CPA = \frac{Cost \ $} {Impressions \ \times \ CTR \ \times \ CR}$
#### Với
## $ CTR = \frac{Clicks}{Impressions}$ & $ CR = (\frac{Conversions}{Clicks})$ 

#### Học viên thực hiện các yêu cầu sau, và đối chiếu với kết quả cho trước:

In [2]:
import numpy as np
import warnings
warnings.filterwarnings('ignore')
np.set_printoptions(suppress=True)

### Yêu cầu 1: Cho tập tin cost_conversions.txt chứa thông tin cost và conversions.

In [3]:
# Câu 1: Đọc dữ liệu từ tập tin cost_conversions.txt, và lưu vào numpy array cost_conversions chứa thông tin cost và conversions của các mẫu
# Gợi ý: dùng np.loadtxt('ten_tap_tin', dtype=kieu_du_lieu)
cost_conversions = np.loadtxt('cost_conversions.txt',dtype=np.float)


In [4]:
# Câu 2: Cho biết numpy array cost_conversions chứa bao nhiêu mẫu. 
cost_conversions.shape

(1143, 2)

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(1143, 2)</pre>
  
</details>

In [5]:
# Hiển thị 5 mẫu đầu tiên của array cost_conversions
cost_conversions[0:5]

array([[1.43, 2.  ],
       [1.82, 2.  ],
       [0.  , 1.  ],
       [1.25, 1.  ],
       [1.29, 1.  ]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([[1.43, 2.  ],
       [1.82, 2.  ],
       [0.  , 1.  ],
       [1.25, 1.  ],
       [1.29, 1.  ]])</pre>
  
</details>

In [6]:
# Câu 3: Cho biết mẫu nào có cost cao nhất.
np.amax(cost_conversions,axis=0,keepdims=True)

array([[639.95,  60.  ]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Mẫu có cost cao nhất:  [[639.95  60.  ]]
</pre>
  
</details>

In [7]:
# Câu 4: Cho biết mẫu nào có conversion cao nhất
np.amax(cost_conversions,axis=0,keepdims=True)

array([[639.95,  60.  ]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Mẫu có conversion cao nhất:  [[639.95  60.  ]]
</pre>
  
</details>

In [11]:
# Câu 5: Cho biết giá trị trung bình, trung vị của cost, convsersions
print('Giá trị trung bình của cost:',np.mean(cost_conversions[:,0]))
print('Giá trị trung vị của cost:',np.median(cost_conversions[:,0]))
print('Giá trị trung bình của conversions:',np.mean(cost_conversions[:,1]))
print('Giá trị trung vị của conversions:',np.median(cost_conversions[:,1]))

Giá trị trung bình của cost: 51.360656167979
Giá trị trung vị của cost: 12.37
Giá trị trung bình của conversions: 2.8556430446194225
Giá trị trung vị của conversions: 1.0


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị trung bình của cost:  51.360656167979
Giá trị trung vị của cost:  12.37
Giá trị trung bình của convsersions:  2.8556430446194225
Giá trị trung vị của convsersions:  1.0
</pre>
  
</details>

In [12]:
# Câu 6: Tạo array CPA và tính theo công thức CPA = (COST/CONVERSIONS). Hiển thị 5 mẫu đầu tiên của array CPA
cpa = cost_conversions[:,0] / cost_conversions[:,1]
cpa[:5]

array([0.715, 0.91 , 0.   , 1.25 , 1.29 ])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([0.715, 0.91 , 0.   , 1.25 , 1.29 ])</pre>
  
</details>

In [36]:
# Câu 7: Tạo numpy array cost_conversions_CPA, lấy thông tin từ cost_conversions và CPA, 
# sao cho mỗi mẫu sẽ chứa 3 thông tin [cost, conversions, CPA]
cost_conversions_CPA = np.insert(cost_conversions,2,cpa,axis=1)
cost_conversions_CPA[:5]


array([[1.43 , 2.   , 0.715],
       [1.82 , 2.   , 0.91 ],
       [0.   , 1.   , 0.   ],
       [1.25 , 1.   , 1.25 ],
       [1.29 , 1.   , 1.29 ]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([[1.43 , 2.   , 0.715],
       [1.82 , 2.   , 0.91 ],
       [0.   , 1.   , 0.   ],
       [1.25 , 1.   , 1.25 ],
       [1.29 , 1.   , 1.29 ]])</pre>
  
</details>

In [43]:
# Câu 8: Xóa tất cả các mẫu có conversions = 0 hoặc cost = 0 trong cost_conversions_CPA. 
print('Trước khi xóa cost_conversions_CPA có',cost_conversions_CPA.shape,'  mẫu')
condition = np.where((cost_conversions_CPA[:,0] == 0) | (cost_conversions_CPA[:,1] == 0))
cost_conversions_CPA = np.delete(cost_conversions_CPA, condition, axis=0)
# Cho biết lúc này cost_conversions_CPA còn bao nhiêu mẫu?
print('Sau khi xóa cost_conversions_CPA có',cost_conversions_CPA.shape,'  mẫu')

Trước khi xóa cost_conversions_CPA có (1143, 2)   mẫu
Sau khi xóa cost_conversions_CPA có (931, 2)   mẫu


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Trước khi xóa, cost_conversions_CPA có (1143, 3) mẫu.
Sau khi xóa, cost_conversions_CPA có (931, 3) mẫu.
</pre>
  
</details>

### Yêu cầu 2: Cho tập tin f_data_new.txt chứa thông tin cost, clicks, impressions, CR.

In [44]:
# Câu 1: Đọc dữ liệu từ tập tin f_data_new.txt, và lưu vào numpy array f_data chứa thông tin cost, clicks, impressions, CR của các mẫu.
# Gợi ý: dùng np.loadtxt('ten_tap_tin', dtype=kieu_du_lieu)
f_data = np.loadtxt('f_data_new.txt', dtype=float)

In [45]:
# Câu 2: Cho biết f_data chứa bao nhiêu mẫu. 
f_data.shape

(936, 4)

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(936, 4)</pre>
  
</details>

In [46]:
# Hiển thị 5 mẫu đầu tiên của f_data
f_data[:5]

array([[    1.43,     1.  ,  7350.  ,     2.  ],
       [    1.82,     2.  , 17861.  ,     1.  ],
       [    1.25,     1.  ,  4259.  ,     1.  ],
       [    1.29,     1.  ,  4133.  ,     1.  ],
       [    4.77,     3.  , 15615.  ,     0.33]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([[    1.43,     1.  ,  7350.  ,     2.  ],
       [    1.82,     2.  , 17861.  ,     1.  ],
       [    1.25,     1.  ,  4259.  ,     1.  ],
       [    1.29,     1.  ,  4133.  ,     1.  ],
       [    4.77,     3.  , 15615.  ,     0.33]])</pre>
  
</details>

In [47]:
# Câu 3: Xóa các mẫu có impressions = 0 hoặc CR = 0 của f_data. 
condition = np.where((f_data[:,2] == 0) | (f_data[:,3] == 0))
f_data = np.delete(f_data,condition,axis=0)
# Cho biết lúc này, f_data còn bao nhiêu mẫu?
print("Số mẫu f_data còn lại:",f_data.shape)

Số mẫu f_data còn lại: (929, 4)


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(929, 4)</pre>
  
</details>

In [48]:
# Câu 4: Tạo numpy array CTR và tính theo công thức nêu trên. 
ctr = f_data[:,1] / f_data[:,2]
# Hiển thị 5 mẫu đầu tiên của array CTR.
ctr[:5]

array([0.00013605, 0.00011198, 0.0002348 , 0.00024195, 0.00019212])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([0.00013605, 0.00011198, 0.0002348 , 0.00024195, 0.00019212])</pre>
  
</details>

In [49]:
# Câu 5: Tạo numpy array CPA và tính theo công thức cách 2. 
cpa = f_data[:,0] / (f_data[:,2]*ctr[:]*f_data[:,3])
# Hiển thị 5 mẫu đầu tiên của array CPA.
cpa[:5]

array([0.715     , 0.91      , 1.25      , 1.29      , 4.81818182])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([0.715     , 0.91      , 1.25      , 1.29      , 4.81818182])</pre>
  
</details>

In [51]:
# Câu 6: Cho biết giá trị CPA lớn nhất, nhỏ nhất, trung bình và trung vị.
print('CPA lớn nhất %f'%np.max(cpa))
print('CPA nhỏ nhất %f'%np.min(cpa))
print('CPA trung bình %f'%np.mean(cpa))
print('CPA trung vị %f'%np.median(cpa))

CPA lớn nhất 158.087719
CPA nhỏ nhất 0.180000
CPA trung bình 19.365150
CPA trung vị 11.807179


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị lớn nhất của CPA:  158.0877192982456
Giá trị nhỏ nhất của CPA:  0.18
Giá trị trung bình của CPA:  19.365149795676988
Giá trị trung vị của CPA:  11.807179487179488
</pre>
  
</details>

In [67]:
# Câu 7: Tạo một numpy array f_all,
# lấy thông tin từ f_data, CTR, CPA, sao cho mỗi mẫu sẽ chứa các thông tin cost, clicks, impressions, CR, CTR, CPA.
f_all = np.insert(f_data,4,ctr,axis=1)
f_all = np.insert(f_all,5,cpa,axis=1)

f_all

array([[     1.43      ,      1.        ,   7350.        ,
             2.        ,      0.00013605,      0.715     ],
       [     1.82      ,      2.        ,  17861.        ,
             1.        ,      0.00011198,      0.91      ],
       [     1.25      ,      1.        ,   4259.        ,
             1.        ,      0.0002348 ,      1.25      ],
       ...,
       [    40.29      ,     28.        , 151531.        ,
             0.07      ,      0.00018478,     20.55612245],
       [   198.71      ,    135.        , 790253.        ,
             0.06      ,      0.00017083,     24.53209877],
       [   165.61      ,    114.        , 513161.        ,
             0.04      ,      0.00022215,     36.31798246]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([[    1.43      ,     1.        ,  7350.        ,     2.        ,
            0.00013605,     0.715     ],
       [    1.82      ,     2.        , 17861.        ,     1.        ,
            0.00011198,     0.91      ],
       [    1.25      ,     1.        ,  4259.        ,     1.        ,
            0.0002348 ,     1.25      ],
       [    1.29      ,     1.        ,  4133.        ,     1.        ,
            0.00024195,     1.29      ],
       [    4.77      ,     3.        , 15615.        ,     0.33      ,
            0.00019212,     4.81818182]])</pre>
  
</details>

In [69]:
# Câu 8: Lưu kết quả vào tập tin f_all.txt
np.savetxt('f_all.txt',f_all)